File: iteration2/code/src/coursetracking/utils/ControlCenter.java
package coursetracking.utils;


import java.io.File;
import java.io.FileWriter;
import java.io.FilenameFilter;
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

import coursetracking.models.Student;
import coursetracking.models.TakenCourse;
import coursetracking.models.Transcript;
import coursetracking.models.Config;
import coursetracking.models.Course;
import coursetracking.models.Elective;


import com.google.gson.Gson;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;


public class ControlCenter {
	
	    private Config config;
	    private Data data;
	    private Gson gson = Utils.getInstance().getGson();
	    private ArrayList<Student> students = new ArrayList<>();
	    private FilenameFilter transcriptFilter = new FilenameFilter() {

	        @Override
	        public boolean accept(File dir, String name) {
	            return name.startsWith("150") && name.endsWith(".json");
	        }

	    };

	   public void start() throws Exception {
	        readInput();
	        readNames();
	        if (Utils.getInstance().getOutputPath().listFiles(transcriptFilter).length == 0 || config.studentGeneration)
	            createStudents();
	        else {
	            readStudents();
	            gradesOfLastSemester();
	            newTermRegistration();
	        }
	        writeOutputJson();
	    }

	    // assigns grades to the last registered (non-graded yet) courses before new
	    // registration
	    public boolean gradesOfLastSemester() {
	        for (Student st : data.students) {
	            int numberOfTrc = st.getTranscripts().size();
	            // if the term parameter isn't changed, no grades will be assigned
	            if ((config.registrationTerm.toLowerCase().equals("fall") && numberOfTrc % 2 != 0 && numberOfTrc!=8) ||
	                    (config.registrationTerm.toLowerCase().equals("spring") && numberOfTrc % 2 == 0 && numberOfTrc!=8))
	                return false;

	            Transcript trscript = st.getTranscripts().get(numberOfTrc - 1); // last semesterCourses in transcript
	            for (TakenCourse tkc : trscript.getSemesterCourses()) {
	                int rnd = random.nextInt(9);
	                tkc.setLetterGrade(letterNotes[rnd]);
	            }
	            trscript.calculate();
	            st.calculate();
	            st.save();
	        }
	        return true;
	    }

	    // students register new term (term parameter changed in input.json)
	    public boolean newTermRegistration() {
	        //setElectivesNull();//Electives new term
	        for (Student student : data.students) {
	            int lastRegisteredSem = student.getTranscripts().size();
	            int newSem = lastRegisteredSem + 1;

	            // if the term parameter isn't changed, no new registration will be done
	            if ((config.registrationTerm.toLowerCase().equals("fall") && lastRegisteredSem % 2 != 0 && lastRegisteredSem!=8 ) ||
	                    (config.registrationTerm.toLowerCase().equals("spring") && lastRegisteredSem % 2 == 0 && lastRegisteredSem!=8))
	                return false;

	            if (lastRegisteredSem == 8)
	                continue; // (for now) there will be no new registration for students who took all courses

	            Transcript trscript = new Transcript();
	            for (Course c : config.curriculum)   {
	                Boolean quota = true;
	                Random rand = new Random();
	                if (newSem == c.getSemester() && student.canTakeCourse(c)) {
	                    quota = true;
	                    //Elective Part Starts
	                    if(c.getType() != null){//checks for elective course
	                        for( Elective e: config.electives ){
	                            if( e.type.equals(c.getType()) ){
	                            	int electiveRandom = rand.nextInt(getElectiveQuota(e.type));
	                                while( student.isStudentEnrolled(e.courses.get(electiveRandom)) ){
	                                    electiveRandom = rand.nextInt(getElectiveQuota(e.type));//creates for random course inside electives 
	                                }
	                                if(e.isQuotaFull(e.courses.get(electiveRandom))){
										e.setSemester(e.courses.get(electiveRandom), c.getSemester());
	                                    e.courses.get(electiveRandom).addQuotaProblem(student);
	                                    student.addFeedbackQuota(e.courses.get(electiveRandom));
	                                    quota = false;
	                                } 
	                                else {
	                                    c = e.courses.get(electiveRandom);
	                                    c.setType(e.type);
	                                }
	                            }
	                        }
	                    }
	                    //Elective Part Ends
	                    TakenCourse tc;
	                    if(quota){
	                        tc = new TakenCourse(c); // newly registered courses has no grade attribute
	                        trscript.addCourse(tc);
	                        c.enrollStudent(student);
	                    }
	                }
	            }
	            trscript.setSemester(newSem);
	            trscript.calculate();
	            student.addTranscript(trscript);
	            student.calculate();
	            student.save();
	            
	        }
	        return true;
	    }
	    public void setElectivesNull(){
	        for(Elective e: config.electives){
	            e.setNewTerm();
	        }
	    }

	    public void readStudents() throws Exception {
	        data.students = new ArrayList<>();
	        for (File f : Utils.getInstance().getOutputPath().listFiles(transcriptFilter)) {
	            if (f.getName().endsWith(".json")) {
	                String data = "";
	                Scanner myReader = new Scanner(f);
	                while (myReader.hasNextLine()) {
	                    data += myReader.nextLine();
	                }
	                myReader.close();
	                Student st = gson.fromJson(data, Student.class);
	                st.calculate();
	                this.data.students.add(st);
	            }
	        }
	    }

	    public void readNames() throws Exception {
	        File input = Utils.getInstance().getResource("allStudentNames.json");
	        String data = "";
	        Scanner myReader = new Scanner(input);
	        while (myReader.hasNextLine()) {
	            data += myReader.nextLine();
	        }
	        myReader.close();
	        this.data = gson.fromJson(data, Data.class);
	    }

	    public void readInput() throws Exception {
	        File input = Utils.getInstance().getResource("input.json");
	        String data = "";
	        Scanner myReader = new Scanner(input);
	        while (myReader.hasNextLine()) {
	            data += myReader.nextLine();
	        }
	        myReader.close();
	        config = gson.fromJson(data, Config.class);
	    }

	    public void createStudents() {// Creates Students
	        int num = 150121001, index = 0;
	        int sem = (config.registrationTerm.toLowerCase().equals("spring")) ? 2 : 1;
	        for (; sem <= 8; sem += 2) {
	            for (int id = num; id < num + 70; id++) {
	                Student std = new Student(data.students.get(index));
	                std.setId(id);
	                getCoursesBySemester(sem, std);
	                students.add(std);
	                std.save();
	                index++;
	            }
	            num = num - 1000;
	        }
	    }

	    private Random random = new Random(54364564);
	    private String[] letterNotes = { "AA", "BA", "BB", "CB", "CC", "DC", "DD", "FD", "FF" };

	    void getCoursesBySemester(int semester, Student st) {
	        TakenCourse tc;
	        for (int i = 1; i <= semester; i++) {
	            Transcript trscript = new Transcript();
	            for (Course c : config.curriculum) {
	                //Checks for semester
	                if (i == c.getSemester()) {
	                	int rnd = random.nextInt(9);
	                    boolean newEnrollment = false;
	                    boolean quota = true;
						//Elective Part Starts
						if(c.getType() != null){//checks for elective course
							Random rand = new Random();
							for( Elective e: config.electives ){
								if( e.type.equals(c.getType()) ){
									int electiveRandom = rand.nextInt(getElectiveQuota(c.getType()));//creates for random course inside electives
									while( st.isStudentEnrolled(e.courses.get(electiveRandom)) ){
										electiveRandom = rand.nextInt(getElectiveQuota(e.type));//creates for random course inside electives 
									}
									e.setSemester(e.courses.get(electiveRandom), c.getSemester());
									if(e.isQuotaFull(e.courses.get(electiveRandom))){
										e.courses.get(electiveRandom).addQuotaProblem(st);
										st.addFeedbackQuota(e.courses.get(electiveRandom));
										quota = false;
									} 
									else c = e.courses.get(electiveRandom);
								}
							}
						}
						//Elective Part Ends
	                	if (i == semester){
							tc = new TakenCourse(c); // newly registered courses has no grade attribute
	                        newEnrollment = true;
	                    }
						else
							tc = new TakenCourse(c, letterNotes[rnd]);

	                    if (st.canTakeCourse(c))
	                    {
	                        if(quota) trscript.addCourse(tc);
	                        if(newEnrollment && quota) c.enrollStudent(st);
	                    }
	                }
	            }
	            trscript.setSemester(i);
	            trscript.calculate();
	            st.addTranscript(trscript);
	        }
	        st.calculate();
	    }

	    public int getElectiveQuota(String type){
	        int ret = 0;
	        for(Elective e: config.electives){
	            if(type.equals(e.type)){
	                ret = e.getElectiveQuantity();
	            }
	        }
	        return ret;
	    }

	    // writes a general feedback for all students on the screen and into OUTPUT.json
	    // file under output folder
	    public void writeOutputJson() {
	        Output outputObj = new Output();
	        for (Course c : config.curriculum) {
	            for (String s : c.getFeedback()) {
	                System.out.println(s);
	                outputObj.addFeedback(s);
	            }
	        }
	        for(Elective e: config.electives)
	        {
	            for(Course c: e.courses){
	                for(String s: c.getFeedback()){
	                    System.out.println(s);
	                    outputObj.addFeedback(s);
	                }
	            }
	        }
	        try {
	            File output = new File(Utils.getInstance().getOutputPath(), "OUTPUT.json");
	            if (!output.exists())
	                output.createNewFile();
	            FileWriter writer = new FileWriter(output, false);
	            writer.write(gson.toJson(outputObj));
	            writer.close();
	        } catch (Exception e) {
	            System.out.println(e);
	        }
	    }
	}

	class Output {
	    @SerializedName("feedback")
	    @Expose
	    ArrayList<String> feedbackList;

	    public void addFeedback(String feedback) {
	        if (feedbackList == null)
	            feedbackList = new ArrayList<String>();
	        feedbackList.add(feedback);
	    }
	}

	class Data {
	    @SerializedName("fullnames")
	    @Expose
	    public ArrayList<Student> students;
	}
	

File: iteration3/src/model/ControlCenter.py

class ControlCenter:

    def __init__(self):
        
        self.__config = None
        self.__data = None
        self.__gson = Utils.getInstance().getGson()
        self.__students = []
        self.__transcriptFilter = None
        self.__letterNotes = ["AA", "BA", "BB", "CB", "CC", "DC", "DD", "FD", "FF"]


    def start(self):
        self.readInput()
        self.readNames()
        if Utils.getInstance().getOutputPath().listFiles(self.__transcriptFilter).length == 0 or self.__config.studentGeneration:
            self.createStudents()
        else:
            self.readStudents()
            self.gradesOfLastSemester()
            self.newTermRegistration()
        self.writeOutputJson()

    class FilenameFilterAnonymousInnerClass():

        def accept(self, dir, name):
            return name.startswith("150") and name.endswith(".json")


    # assigns grades to the last registered (non-graded yet) courses before new
    # registration
    def gradesOfLastSemester(self):
        for st in self.__data.students:
            numberOfTrc = len(st.getTranscripts())
            # if the term parameter isn't changed, no grades will be assigned
            if (self.__config.registrationTerm.casefold() == "fall" and math.fmod(numberOfTrc, 2) != 0 and numberOfTrc!=8) or (self.__config.registrationTerm.casefold() == "spring" and math.fmod(numberOfTrc, 2) == 0 and numberOfTrc!=8):
                return False

            trscript = st.getTranscripts()[numberOfTrc - 1] # last semesterCourses in transcript
            for tkc in trscript.getSemesterCourses():
                rnd = self.__random.nextInt(9)
                tkc.setLetterGrade(self.__letterNotes[rnd])
            trscript.calculate()
            st.calculate()
            st.save()
        return True

    # students register new term (term parameter changed in input.json)
    def newTermRegistration(self):
        #setElectivesNull();//Electives new term
        for student in self.__data.students:
            lastRegisteredSem = len(student.getTranscripts())
            newSem = lastRegisteredSem + 1

            # if the term parameter isn't changed, no new registration will be done
            if (self.__config.registrationTerm.casefold() == "fall" and math.fmod(lastRegisteredSem, 2) != 0 and lastRegisteredSem!=8) or (self.__config.registrationTerm.casefold() == "spring" and math.fmod(lastRegisteredSem, 2) == 0 and lastRegisteredSem!=8):
                return False

            if lastRegisteredSem == 8:
                continue # (for now) there will be no new registration for students who took all courses

            trscript = coursetracking.models.Transcript()
            for c in self.__config.curriculum:
                quota = True
                rand = java.util.Random()
                if newSem == c.getSemester() and student.canTakeCourse(c):
                    quota = True
                    #Elective Part Starts
                    if c.getType() is not None:
                        for e in self.__config.electives:
                            if e.type == c.getType():
                                electiveRandom = rand.nextInt(self.getElectiveQuota(e.type))
                                while student.isStudentEnrolled(e.courses[electiveRandom]):
                                    electiveRandom = rand.nextInt(self.getElectiveQuota(e.type)) #creates for random course inside electives
                                if e.isQuotaFull(e.courses[electiveRandom]):
                                    e.setSemester(e.courses[electiveRandom], c.getSemester())
                                    e.courses[electiveRandom].addQuotaProblem(student)
                                    student.addFeedbackQuota(e.courses[electiveRandom])
                                    quota = False
                                else:
                                    c = e.courses[electiveRandom]
                                    c.setType(e.type)
                    #Elective Part Ends
                    tc = None
                    if quota:
                        tc = coursetracking.models.TakenCourse(c) # newly registered courses has no grade attribute
                        trscript.addCourse(tc)
                        c.enrollStudent(student)
            trscript.setSemester(newSem)
            trscript.calculate()
            student.addTranscript(trscript)
            student.calculate()
            student.save()

        return True
    def setElectivesNull(self):
        for e in self.__config.electives:
            e.setNewTerm()

    def readStudents(self):
        self.__data.students = []
        for f in Utils.getInstance().getOutputPath().listFiles(self.__transcriptFilter):
            if f.getName().endsWith(".json"):
                data = ""
                myReader = java.util.Scanner(f)
                while myReader.hasNextLine():
                    data += myReader.nextLine()
                myReader.close()
                st = self.__gson.fromJson(data, coursetracking.models.Student)
                st.calculate()
                self.__data.students.append(st)

    def readNames(self):
        input = Utils.getInstance().getResource("allStudentNames.json")
        data = ""
        myReader = Scanner(input)
        while myReader.hasNextLine():
            data += myReader.nextLine()
        myReader.close()
        self.data = gson.fromJson(data, Data)



    def readInput(self):
        input = Utils.getInstance().getResource("input.json")
        data = ""
        myReader = Scanner(input)
        while myReader.hasNextLine():
            data += myReader.nextLine()
        myReader.close()
        config = gson.fromJson(data, Config)

    def createStudents(self):
        num = 150121001
        index = 0
        sem = 2 if (config.registrationTerm.toLowerCase() == "spring") else 1
        while sem <= 8:
            id = num
            while id < num + 70:
                std = Student(data.students.get(index))
                std.setId(id)
                self.getCoursesBySemester(sem, std)
                students.add(std)
                std.save()
                index += 1
                id += 1
            num = num - 1000
            sem += 2

    def getCoursesBySemester(self, semester, st):
        tc = None
        for i in range(1, semester + 1):
            trscript = Transcript()
            for c in config.curriculum:
                # Checks for semester
                if i == c.getSemester():
                    rnd = self.__random.nextInt(9)
                    newEnrollment = False
                    quota = True
                    # Elective Part Starts
                    if c.getType() is not None:
                        rand = Random()
                        for e in config.electives:
                            if e.type is c.getType():
                                electiveRandom = rand.nextInt(
                                    self.getElectiveQuota(c.getType()))  # creates for random course inside electives
                                while st.isStudentEnrolled(e.courses.get(electiveRandom)):
                                    electiveRandom = rand.nextInt(
                                        self.getElectiveQuota(e.type))  # creates for random course inside electives
                                e.setSemester(e.courses.get(electiveRandom), c.getSemester())
                                if e.isQuotaFull(e.courses.get(electiveRandom)):
                                    e.courses.get(electiveRandom).addQuotaProblem(st)
                                    st.addFeedbackQuota(e.courses.get(electiveRandom))
                                    quota = False
                                else:
                                    c = e.courses.get(electiveRandom)
                    # Elective Part Ends
                    if i == semester:
                        tc = TakenCourse(c)  # newly registered courses has no grade attribute
                        newEnrollment = True
                    else:
                        tc = TakenCourse(c, self.__letterNotes[rnd])

                    if st.canTakeCourse(c):
                        if quota:
                            trscript.addCourse(tc)
                        if newEnrollment and quota:
                            c.enrollStudent(st)
            trscript.setSemester(i)
            trscript.calculate()
            st.addTranscript(trscript)
        st.calculate()

    def getElectiveQuota(self, type):
        ret = 0
        for e in config.electives:
            if type == e.type:
                ret = e.getElectiveQuantity()
        return ret

    # writes a general feedback for all students on the screen and into OUTPUT.json
    # file under output folder
    def writeOutputJson(self):
        outputObj = Output()
        for c in config.curriculum:
            for s in c.getFeedback():
                print(s)
                outputObj.addFeedback(s)
        for e in config.electives:
            for c in e.courses:
                for s in c.getFeedback():
                    print(s)
                    outputObj.addFeedback(s)
        try:
            output = File(Utils.getInstance().getOutputPath(), "OUTPUT.json")
            if not output.exists():
                output.createNewFile()
            writer = FileWriter(output, False)
            writer.write(gson.toJson(outputObj))
            writer.close()
        except Exception as e:
            print(e)



    class Output:

        def __init__(self):
            self.feedbackList = None

        def addFeedback(self, feedback):
            if self.feedbackList is None:
                self.feedbackList = []
            self.feedbackList.append(feedback)

    class Data:
        def __init__(self):
            self.students = None


